Linear Exchange Matrix Build (LinK) in Direct SCF algorithm (#2359)
psi4/psi4: doc/sphinxman/source/scf.rst

basis set defined for all atoms in the system, or set |scf__df_scf_guess|
to false, which disables this acceleration entirely.

Linear-Scaling Exchange Build
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Methods that build K matrices in linear time have been developed. One of which is the Linear Exchange (LinK) Method 
> to build K matrices, described in [Ochsenfeld:1998:1663]_. The LinK method is available to use as part
> of the Direct SCF algorithm. LinK offers a fast alternative to density-fitted K builds when the amount of available memory is low, 
> especially when combined with Density-matrix based ERI screening and incremental Fock builds.

Reviewer: ```suggestion
Methods that build exchange matrices in linear time with the size of the system (linear-time K builds)  are often useful for large calculations.
The Linear Exchange (LinK) Method, described in [Ochsenfeld:1998:1663]_, is available to use in the Direct SCF algorithm.
LinK is a competive alternative to density-fitted K builds with small orbital basis sets that lead to considerable sparsity in large systems, and/or when there is not enough memory to perform the density-fitting procedure in-core.
LinK is especially powerful when combined with density-matrix based ERI screening and incremental Fock builds which decrease the number of significant two-electron integrals to calculate.
```

> LinK is a competive alternative to density-fitted K builds with small orbital basis sets that lead to considerable sparsity in large systems, and/or when there is not enough memory to perform the density-fitting procedure in-core.

Author: Ah, this should be reworded to non-diffuse basis sets. LinK would still work with tz and qz as long as there are not too many diffuse basis functions.

> LinK is a competitive alternative to density-fitted K builds with non-diffuse orbital basis sets that lead to considerable sparsity in large systems, and/or when there is not enough memory to perform the density-fitting procedure in-core.

Author: Updated comments. LinK is still technically linear scaling with diffuse functions, but the linear-scaling behavior is not observed until much bigger systems due to slower decaying overlaps.

To turn on and control the LinK algorithm, here are the list of options provided.

  |scf__do_linK|: Defaults to false. If turned on, the K matrix will be built using the algorithm described in [Ochsenfeld:1998:1663]_.

  |scf__linK_ints_tolerance|: Defaults to 1.0e-12. The integral screening tolerance used for sparsity-prep in the LinK algorithm. Note that

>     this option is separate from and independent of the |scf__ints_tolerance| option. A tighter value is recommended for more diffuse basis sets (1.0e-14).

Reviewer: Maybe a stupid question, but why is this different from `scf_ints_tolerance`?

Author: I made it separate in case we want to control the error of LinK separately from the error of Schwarz or Density Screening. This is an expert option and should default to whatever the value of `scf__ints_tolerance` is.

>   |scf__linK_ints_tolerance|: Defaults to 1.0e-12. The integral screening tolerance used for sparsity-prep in the LinK algorithm. Note that

Reviewer 2: I agree with Jonathon on this point.

basis set defined for all atoms in the system, or set |scf__df_scf_guess|
to false, which disables this acceleration entirely.

Linear-Scaling Fock Builds
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Methods that build J and K matrices in linear time have been developed. Among these include

Author: Got rid of CFMM references

the Continuous Fast Multipole Method (CFMM) method to build J matrices, described in [White:1994:8]_,
as well as the Linear Exchange (LinK) Method to build K matrices, described in [Ochsenfeld:1998:1663]_.
Currently, the CFMM method has not been implemented yet, but the LinK method is available to use as part
of the Direct SCF algorithm. Linear-Scaling direct algorithms are a fast alternative to density
fitting when the amount of available memory is low, especially when combined with Density-matrix based ERI
screening and incremental Fock builds.

To turn on and control the LinK algorithm, here are the list of options provided.

> DO_LINEAR_EXCHANGE

Reviewer 3: Follow the style of the CPHF options mentioned on this page (`|cphf__solver_maxiter|: maximum number of iterations`) for these options.

    Defaults to false. If turned on, the K matrix will be built using the linear scaling algorithm described in
    [Ochsenfeld:1998:1663]_.

LINK_INTS_TOLERANCE
    Defaults to 1.0e-12. The integral screening tolerance used for sparsity-prep in the LinK algorithm. Note that

>     this option is separate from and independent of the INTS_TOLERANCE option. A tighter value is reccomended for

Reviewer 3: ```suggestion
    this option is separate from and independent of the INTS_TOLERANCE option. A tighter value is recommended for
```

Can you give a rough heuristic for deciding how tight of a value you need for diffuse basis sets? You can't go much lower without hitting machine precision.

basis set defined for all atoms in the system, or set |scf__df_scf_guess|
to false, which disables this acceleration entirely.

> Linear-Scaling Exchange Build
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> Methods that build exchange matrices in linear time with the size of the system (linear-time K builds) are often useful for large calculations.
> The Linear Exchange (LinK) Method, described in [Ochsenfeld:1998:1663]_, is available to use in the Direct SCF algorithm.
> LinK exploits shell pair sparsity in the density matrix, which when combined with the overlap sparsity between shell pairs, leads to linear-scaling K builds.
> LinK is most competitive when used with non-diffuse orbital basis sets, since orbital and density overlaps decay slower with diffuse functions.
> LinK is especially powerful when combined with density-matrix based ERI screening and incremental Fock builds, which decrease the number of significant two-electron integrals to calculate.

Reviewer 4: ```suggestion
LinK Exchange
~~~~~~~~~~~~~~

Large SCF calculations can benefit from specialized screening procedures that further reduce the scaling of the ERI contribution to the Fock matrix.
LinK, the linear-scaling exchange method described in [Ochsenfeld:1998:1663]_, is available with the direct SCF algorithm (|globals__scf_type| set to ``DIRECT``).
LinK achieves linear-scaling by exploiting shell pair sparsity in the density matrix and overlap sparsity between shell pairs.
This method is most competitive when used with non-diffuse orbital basis sets, since orbital and density overlaps decay slower with diffuse functions.
LinK is especially powerful when combined with density-matrix based ERI screening (set |globals__screening| to ``DENSITY``) and incremental Fock builds (set |scf__incfock| to ``TRUE``), which decrease the number of significant two-electron integrals to calculate.
```

> NOTE: Turning on LinK is currently only recommended for research and development purposes, and not for performance,

Reviewer 4: I prefer this wording:
```suggestion
The LinK code is currently under development and should not be used.
```

Reviewer 4: This note should be at the top of the LinK section. Also, instead of writing out the note as plaintext, you can use the `warning` feature in sphinx. See the sapt.rst source / sapt.html doc page.

> since a fast J matrix build compatible to use with LinK has not been developed yet.

Author: @zachglick @JonathonMisiewicz How does this sound for explaining the current state of the LinK code to users?

basis set defined for all atoms in the system, or set |scf__df_scf_guess|
to false, which disables this acceleration entirely.

LinK Exchange

> ~~~~~~~~~~~~~~

Reviewer 4: Oops, my bad
```suggestion
~~~~~~~~~~~~~
```
